### DISCUSSION TITLE
Should we test tiddlywiki for valid HTML?

### DISCUSSION THREAD

**Discussion by BurningTreeC** [TiddlyWiki/TiddlyWiki5#6662] [Q&A]
Hello all,

I found [this site](https://matthiasott.com/notes/debugging-html-with-advanced-css-selectors) with information  about valid HTML

It points to [this Google Chrome extension](https://chrome.google.com/webstore/detail/alix-for-chrome/aepmadgjacfjcneccddiccnkbpimobge) which outlines all problematic elements on a page

I tried the extension on tiddlywiki.com and found some problems but I've also seen that we can do something against it

Should we go towards valid HTML in the core and make some adjustments?

Best wishes,
Simon

**Comment by kookma** [comment]
While it seems some issues will not be solved, but I am for fixing those issues you can!

**Comment by pmario** [comment]
I think, that's a good idea and it shouldn't be too difficult to create a CSS configuration and import it as a JSON for testing. I think, it would be interesting, to see, what's going on.

**Comment by Jermolene** [comment]
When we were investigating that Chrome bug related to rendering of block elements inside inline elements I downloaded the W3C validator. It's a horrible process to install it on a Mac, mucking around with Python. Once I'd got it running I quickly realised that we are a really, really long way from producing valid HTML for our standard layout. A big problem is that the HTML spec doesn't actually allow div elements to be nested inside paragraphs, something that we do a lot.

Nonetheless, I would favour bringing in a validator, but perhaps a different one that can be installed via npm more easily. And then we'd have to carefully figure out which parts of the system would make sense to test.

**Reply by pmario** [reply]
>... Once I'd got it running I quickly realised that we are a really, really long way from producing valid HTML for our standard layout. A big problem is that the HTML spec doesn't actually allow div elements to be nested inside paragraphs, something that we do a lot.

I think, there is an outstanding PR, which is 99% there, to remove most of them. We should really try to get that one going again. 

The CSS "trick" may be able to make us aware of some other oversights, if it gives us some visual cues. ..

**Reply by Jermolene** [reply]
> I think, there is an outstanding PR, which is 99% there, to remove most of them. We should really try to get that one going again.

That PR partially addresses the situations where the core itself generates mis-nested elements, but it does not address the numerous cases where the mis-nestings come from the way that the wikitext is written. The intentional design of TiddlyWiki's dropdowns embeds DIVs inside inline elements, so fixing this goes deeper than you might think

**Comment by pmario** [comment]
>... The intentional design of TiddlyWiki's dropdowns embeds DIVs inside inline elements, so fixing this goes deeper than you might think

Maybe. .. But if we would be able to see those problems, they may be much more obvious- I'm sure some of them are also low hanging fruits. ... eg: Combinations like `<span><$reveal >...</$reveal ><div>` where we have a high chance that using an HTML TAG as reveal-param will give us an improved and valid results.

I think we should aim to create valid and a "simpler" DOM structures. I tink, here we have a possibility to "see" some problems. 

We often see in the discussion groups, that theme authors have problems with the UI structure and CSS styling. If we can improve this structure, we should aim for it. ... 

It seems "phase 2" of the "ubertransclusion" will create some breaking changes. So there may be a good chance to fix some troubling UI problems on the way too.
